sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ComplexTypes.scala

```diff
@@ -22,54 +22,34 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.catalyst.rules.Rule
 
 /**
-* push down operations into [[CreateNamedStructLike]].
-*/
-object SimplifyCreateStructOps extends Rule[LogicalPlan] {
-  override def apply(plan: LogicalPlan): LogicalPlan = {
-    plan.transformExpressionsUp {
-      // push down field extraction
+ * Simplify redundant [[CreateNamedStructLike]], [[CreateArray]] and [[CreateMap]] expressions.
+ */
+object SimplifyExtractValueOps extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform { case p =>
+    p.transformExpressionsUp {

```

Comment:
@dongjoon-hyun , is it safe to simplify it for Aggregate?

Reply:
Sorry for late response, @gatorsmile .
These are expression-level optimization rules. If the original expressions exists in `SELECT`, `GROUP BY`, and `HAVING`, those are simplified in the same way together. Do you have any concerning cases?

Reply:
aggregateExpressions are resolved from groupingExpressions using semanticEquals, while referring to names from input.

Reply:
Expression-level optimizer simplifies both `aggregateExpressions` and `groupingExpressions` together. If the target expression exists at somewhere of both sides, the simplified expression also exists at the same locations of both sides. Given that, `semanticEquals` will work for the updated expressions.

Reply:
how about `select struct(a, b).a from t group by struct(a, b)`? We may optimize it to `select a from t group by struct(a, b)`, which is invalid.

Reply:
Oh, right. I missed to consider that kind of cases.

Reply:
Since `map` is not orderable, that happens for `struct` and `array` types.

Reply:
FWIW I think this is a particular example of a more general problem where expression simplification can break the correspondence between a select expression and its grouping equivalent.

Here's a simpler example:

`SELECT (a + b) - a FROM t GROUP BY a + b`

gets me the following:

`org.apache.spark.sql.AnalysisException: expression 't.`b`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.`

Postgres also has this problem, at least in 9.3: `ERROR: column "t.a" must appear in the GROUP BY clause or be used in an aggregate function Position: 23`